File: jafu/src/main/java/org/springframework/fu/jafu/r2dbc/DataR2dbcDsl.java
package org.springframework.fu.jafu.r2dbc;

import org.springframework.boot.autoconfigure.data.r2dbc.R2dbcDataInitializer;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;
import org.springframework.fu.jafu.AbstractDsl;

import java.util.function.Consumer;

public class DataR2dbcDsl extends AbstractDsl {

    private final Consumer<DataR2dbcDsl> dsl;

    public DataR2dbcDsl(Consumer<DataR2dbcDsl> dsl) {
        this.dsl = dsl;
    }

    public static ApplicationContextInitializer<GenericApplicationContext> dataR2dbc(Consumer<DataR2dbcDsl> dsl) {
        return new DataR2dbcDsl(dsl);
    }

    public DataR2dbcDsl r2dbc(Consumer<R2dbcDsl> r2dbcDsl) {
        new R2dbcDsl(r2dbcDsl).initialize(context);
        return this;
    }

    @Override
    public void initialize(GenericApplicationContext context) {
        super.initialize(context);
        dsl.accept(this);
        new R2dbcDataInitializer().initialize(context);
    }
}

File: kofu/src/main/kotlin/org/springframework/fu/kofu/r2dbc/DataR2dbcDsl.kt
package org.springframework.fu.kofu.r2dbc

import org.springframework.boot.autoconfigure.data.r2dbc.R2dbcDataInitializer
import org.springframework.context.support.GenericApplicationContext
import org.springframework.fu.kofu.AbstractDsl
import org.springframework.fu.kofu.ConfigurationDsl

class DataR2dbcDsl(private val init: DataR2dbcDsl.() -> Unit) : AbstractDsl() {

    override fun initialize(context: GenericApplicationContext) {
        super.initialize(context)
        init()
        R2dbcDataInitializer().initialize(context)
    }
}

/**
 * Configure R2DBC support.
 * @see R2dbcDsl
 */
fun ConfigurationDsl.dataR2dbc(r2dbcDsl: DataR2dbcDsl.() -> Unit = {}) {
    DataR2dbcDsl(r2dbcDsl).initialize(context)
}